src/Native/System.Native/pal_mount.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "../config.h"
+#include "pal_mount.h"
+#include <assert.h>
+#include <string.h>
+#include <errno.h>
+
+// Check if we should use getmntinfo or /proc/mounts
+#if HAVE_MNTINFO
+#include <sys/mount.h>
+#else
+#include <sys/statfs.h>
+#include <mntent.h>
+#define STRING_BUFFER_SIZE 8192
+#endif
+
+#if HAVE_MNTINFO
+
+int GetMountInfo(MountPointFound onFound)
+{
+    // getmntinfo returns pointers to OS-internal structs, so we don't need to worry about free'ing the object
+    int result = 0;
+    struct statfs* mounts = nullptr;
+    int count = getmntinfo(&mounts, 0);
+    if (count > 0)
+    {
+        for (int32_t i = 0; i < count; i++)
+        {
+            int32_t nameLength = strlen(mounts[i].f_mntonname);
+            onFound(mounts[i].f_mntonname, nameLength);
+        }
+    }
+    else
+    {
+        // Something went wrong so set our result to errno
+        result = errno;
+    }
+    
+    return result;
+}
+
+#else
+
+int GetMountInfo(MountPointFound onFound)
+{
+    int result = 0;
+    FILE* fp = setmntent("/proc/mounts", MNTOPT_RO);
+    if (fp == nullptr)
+    {
+        result = errno;
+    }
+    else
+    {
+        // The _r version of getmntent needs all buffers to be passed in; however, we don't know how big of a string
+        // buffer we will need, so pick something that seems like it will be big enough.
+        char buffer[STRING_BUFFER_SIZE] = { };
+        mntent entry = { };
+        while (getmntent_r(fp, &entry, buffer, STRING_BUFFER_SIZE) != nullptr)
+        {
+            int32_t nameLength = strlen(entry.mnt_dir);
+            onFound(entry.mnt_dir, nameLength);
+            memset(buffer, 0, STRING_BUFFER_SIZE);
```

Comment:
I don't get the part about null termination. Surely the fields are null-terminated by getmntent_r. There's no other way to tell where one field ends and another begins AFAICT. And zeroing won't help null termination if the whole buffer is used. The structure and man page is much the same as getpwuid_r and we depend on null-termination there.

I don't see how getmntent_r could succeed and we'd find chars from previous calls going from the pointers in the mntent to their terminators unless getmntent_r had a bug.

8KB of memsetting per entry does seem excessive to me.